### PR TITLE
Fix two obscure bugs I ran into while testing

### PR DESCRIPTION
**Pull request by kentonv**

### CODE REVIEW

File: src/workerd/io/io-gate.c++ [hunk 1]
```diff
@@ -62,6 +62,24 @@ kj::Promise<void> InputGate::onBroken() {
...
 
+InputGate::Lock::Lock(InputGate& gate)
+    : gate(&gate),
+      cs(gate.isCriticalSection
+          ? kj::Maybe(kj::addRef(static_cast<CriticalSection&>(gate)))
+          : nullptr) {
+  InputGate* gateToLock = &gate;
+
+  KJ_IF_MAYBE(c, cs) {
+    if (c->get()->state == CriticalSection::REPARENTED) {
```
Comment (bcaimano) [inline src/workerd/io/io-gate.c++:73]: To summarize these ifs to my understanding:
- If our gate is a critical section AND it has been re-parented, get the effective parent.
- If our gate is a critical section AND it has NOT been re-parented, use the gate.
- If our gate is not a critical section, use the gate.

If you're interested, I could suggest some ways to make this code easier to skim. Otherwise, maybe a comment/empty else branch calling out that `cs` is `gate`?

File: src/workerd/io/worker.c++ [hunk 2]
```diff
@@ -3356,20 +3355,41 @@ kj::Promise<void> Worker::Isolate::SubrequestClient::request(
...
       }
     }
 
-    return kj::heap<ResponseStreamWrapper>(kj::atomicAddRef(*constIsolate),
-                                           kj::mv(requestId),
-                                           kj::mv(responseBody),
-                                           encoding,
-                                           *requestMetrics);
+    // Defer to a later turn of the event loop so that it's safe to take a lock.
+    return kj::newPromisedStream(kj::evalLater(
```
Comment (bcaimano) [inline src/workerd/io/worker.c++:3366]: Is there some way we could determine if we had the inspector session without taking the isolate lock? It would be nice if we could avoid the promised stream if we know there is no inspector and I notice the comment about `signalRequest()`. That said, the captures look right to me.

**Review by bcaimano — APPROVED** [review]

File: src/workerd/io/worker.c++ (hunk 2 above)
Comment (kentonv) [reply src/workerd/io/worker.c++:3366]: As mentioned in the comments, in practice we won't reach this code in the first place if there's no inspector session.

I do think this code could use a cleanup but not volunteering for that at the moment.

File: src/workerd/io/io-gate.c++ (hunk 1 above)
Comment (kentonv) [reply src/workerd/io/io-gate.c++:73]: I agree this code is pretty hard to follow. It took me a while to figure it out myself. (And I think I wrote it... though there have been some changes.)

Right now I'm drowning in code reviews so I prefer to merge this as-is and move on.
